File: src/#725/solution.go
package solution

type ListNode struct {
	Val  int
	Next *ListNode
}

func splitListToParts(root *ListNode, k int) []*ListNode {
	res := make([]*ListNode, k)

	if root == nil {
		return res
	}

	current := root
	count := 0
	for current != nil {
		count++
		current = current.Next
	}

	remainder := count % k
	num := count / k //每一條需要幾個
	var prev *ListNode
	current = root

	for i := 0; i < len(res); i++ {
		res[i] = current
		for j := 0; j < num; j++ {
			prev = current
			current = current.Next
		}
		if remainder > 0 {
			prev = current
			current = current.Next
			remainder--
		}
		prev.Next = nil
	}

	return res
}

File: src/#725/solution.js
function ListNode(val, next) {
  this.val = val === undefined ? 0 : val
  this.next = next === undefined ? null : next
}

var splitListToParts = function (root, k) {
  let res = new Array(k).fill(null)
  
  if (root === null) {
    return res
  }

  let current = root
  let count = 0

  while (current !== null) {
    count++
    current = current.next
  }

  let remainder = count % k
  let num = Math.floor(count / k)
  let prev = null
  current = root

  for (let i = 0; i < res.length; i++) {
    res[i] = current
    for (let j = 0; j < num; j++) {
      prev = current
      current = current.next
    }
    if (remainder > 0) {
      prev = current
      current = current.next
      remainder--
    }
    prev.next = null
  }

  return res
}
